### PR TITLE
Adding custom validators

### PR DESCRIPTION
**Pull request by malinthaprasan**
**Added 2 custom validators for validating:**

1. Api-Id - @ValidateApiId
2. Validate equality of parameters - @ValidateParamEquality

**Usage:**

1. Api-id

Please see parameter **apiId**

```Java
    public Response apisApiIdGet(@ApiParam(value = "**API ID** consisting of the name of the API, the identifier of the version and of the provider of the API. \nShould be formatted as **name/version/provider**\n",required=true ) @PathParam("apiId") @ValidateApiId String apiId,
    @ApiParam(value = "Media types acceptable for the response. Should denote XML or JSON, default is JSON."  )@HeaderParam("Accept") String accept,
    @ApiParam(value = "Validator for conditional requests; based on ETag."  )@HeaderParam("If-None-Match") String ifNoneMatch,
    @ApiParam(value = "Validator for conditional requests; based on Last Modified header."  )@HeaderParam("If-Modified-Since") String ifModifiedSince)
    throws NotFoundException {
    return delegate.apisApiIdGet(apiId,accept,ifNoneMatch,ifModifiedSince);
    }
```

Adding a custom error message

```Java
@ValidateApiId(message = "API Id is invalid") 
```

2. Validate equality of parameters

```Java
    @ValidateParamEquality(index0 = 0, index1 = 2)
    public Response applicationsApplicationIdPut(@ApiParam(value = "Application Id",required=true ) @PathParam("applicationId") String applicationId,
    @ApiParam(value = "Application object that needs to be updated" ,required=true ) ApplicationDTO body,
    @ApiParam(value = "Media type of the entity in the request body. Should denote XML or JSON, default is JSON."  )@HeaderParam("Content-Type") String contentType,
    @ApiParam(value = "Validator for conditional requests; based on ETag."  )@HeaderParam("If-Match") String ifMatch,
    @ApiParam(value = "Validator for conditional requests; based on Last Modified header."  )@HeaderParam("If-Unmodified-Since") String ifUnmodifiedSince)
    throws NotFoundException {
    return delegate.applicationsApplicationIdPut(applicationId,body,contentType,ifMatch,ifUnmodifiedSince);
    }
```

2.1 Validate parameter 0 and parameter 2 of the method **applicationsApplicationIdPut** for the equality.

```Java
@ValidateParamEquality(index0 = 0, index1 = 2)
```

2.2 Validate parameter 0 and the field "applicationId" of parameter 1 of the method **applicationsApplicationIdPut** for the equality. i.e Validates the given application id in the request path and the application id in the request body of PUT method.

```Java
@ValidateParamEquality(index0 = 0, index1 = 1, index1Field = "applicationId")
```

2.3 Adding multiple validations for the same method

```Java
    @ValidateParamEquality.List({
            @ValidateParamEquality(index0 = 0, index1 = 1, index1Field = "applicationId"),
            @ValidateParamEquality(index0 = 0, index1 = 2)
    })
    public Response applicationsApplicationIdPut(@ApiParam(value = "Application Id",required=true ) @PathParam("applicationId") String applicationId,
    @ApiParam(value = "Application object that needs to be updated" ,required=true ) ApplicationDTO body,
    @ApiParam(value = "Media type of the entity in the request body. Should denote XML or JSON, default is JSON."  )@HeaderParam("Content-Type") String contentType,
    @ApiParam(value = "Validator for conditional requests; based on ETag."  )@HeaderParam("If-Match") String ifMatch,
    @ApiParam(value = "Validator for conditional requests; based on Last Modified header."  )@HeaderParam("If-Unmodified-Since") String ifUnmodifiedSince)
    throws NotFoundException {
    return delegate.applicationsApplicationIdPut(applicationId,body,contentType,ifMatch,ifUnmodifiedSince);
    }
```

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/validation/ParameterEqualityValidator.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            if (!isValid && ValidateParamEquality.DEFAULT_ERROR_MESSAGE.equals(errorMessage)) {
+                constraintValidatorContext.disableDefaultConstraintViolation();
+                constraintValidatorContext.buildConstraintViolationWithTemplate(
+                        "Values \\{" + firstObj + "," + secondObj + "\\} must be equal").addConstraintViolation();
+            }
+
+            return isValid;
+
+        } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            throw new InternalServerErrorException(e);
```
Comment (nuwand) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/validation/ParameterEqualityValidator.java]: Use proper exception handling.

Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/validation/ParameterEqualityValidator.java]: Fixed in #1528 (https://github.com/wso2/carbon-apimgt/commit/40bb7af6f0d8c412f6e6f80497a52e8d88ff7b9b)

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/validation/ParameterEqualityValidator.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            if (secondField != null && !StringUtils.isEmpty(secondField)) {
+                secondObj = BeanUtils.getProperty(secondObj,
+                        secondField); //if second field is present, take second object as the field value of the secondObj
+            }
+
+            //let @NotNull for checking null values
+            if (firstObj == null || secondObj == null) {
+                return true;
+            }
+
```
Comment (shaniR) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/validation/ParameterEqualityValidator.java]: If the above code segment  (line 66 - 69) is moved before the first if statement, we can get rid of the null check for the subsequent if statements right?

Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/validation/ParameterEqualityValidator.java]: Hi, Please note firstObj and secondObj are initialized in if statements between 56-64

**Pull request merged**